osquery/filesystem/filesystem.cpp

```diff
@@ -276,14 +276,42 @@ static void genGlobs(std::string path,
                      GlobLimits limits) {
   // Use our helped escape/replace for wildcards.
   replaceGlobWildcards(path, limits);
+  // inodes of directory symlinks for loop detection
+  std::set<int> dsym_inos;
 
   // Generate a glob set and recurse for double star.
-  size_t glob_index = 0;
-  while (++glob_index < kMaxRecursiveGlobs) {
+  for (size_t glob_index = 0; ++glob_index < kMaxRecursiveGlobs;) {
     auto glob_results = platformGlob(path);
 
-    for (auto const& result_path : glob_results) {
+    for (auto& result_path : glob_results) {
       results.push_back(result_path);
+      struct stat d_stat;
+
+      if (::lstat(result_path.c_str(), &d_stat) < 0) {
+        continue;
+      }
+
+      if (0 == S_ISDIR(d_stat.st_mode)) {
```

Comment:
Maybe bring this up to the previous `if` comparison and perform both checks in the same conditional? It seems like there might be a way to tighten up this and the other conditionals down below.